### ISSUE TITLE
(nmap) Autohotkey silent install failure

### ISSUE THREAD

**Issue by AdmiringWorm** [chocolatey-community/chocolatey-coreteampackages#1557]
## Expected Behavior

The automated verification test on chocolatey.org to succeed.

## Current Behavior

The verification test for version 7.90 of nmap is currently failing with a message saying that silent installation is only supported in Nmap OEM (I can not find any reference on such a thing on their website).

## Possible Solution

May require the usage of autohotkey (which I have zero to little knowledge about).

## Context

Prevents the latest version of nmap to be approved for download.

## Your Environment

* Package Version used: 7.90
* Operating System and version: Windows Server 2012 R2
* Chocolatey version: 0.10.16-beta
* Install/uninstall gist: https://gist.github.com/choco-bot/da935bb24075c9e8912235a1f48c5624

**Comment by UsernameIsAlreadyTakenByMe** [comment]
I'm not sure they already decided what are they nowadays:
https://github.com/nmap/npcap/blob/master/LICENSE
"Even though Npcap source code is publicly available for review, it is not open source software"
https://nmap.org/
"Nmap ("Network Mapper") is a free and open source (license) utility"

**Comment by nutterthanos** [comment]
Yeah i don't even know how to do silent install now so now what do we do install the oem version or if you said need autohotkey how do we use that to get silent install?

**Comment by RedBaron2** [comment]
@AdmiringWorm 
The best option is to convert the package to use the zip file.
Then using the Chocolatey `Get-UninstallRegistryKey` Helper to find a prior install location, or extract to the `$toolsdir` folder.

**Comment by nutterthanos** [comment]
@RedBaron2 the only thing with that is npcap will need to be a dependency

**Comment by nutterthanos** [comment]
@AdmiringWorm Any Progress?

**Comment by majkinetor** [comment]
OEM installer is commercial one so silent options can not be used anymore. I tried with AHK installer and it seems to work. Zip is also an option but it doesn't have everything setup has (stuff could be added via dependencies).

I think it should work now but will be available from next version as I didn't manage to force update now.

**Comment by majkinetor** [comment]
OK, I managed to publish new version.

Give it a go: https://chocolatey.org/packages/nmap/7.91.0.20201123

**Comment by majkinetor** [comment]
It still fails verifier but for different reason. It looks like it blocked when installing vcredist2008. Its probable that some popup blocked it that showed on old Windows Server that verifier is using. If I can get that popup I could fix it, but I don't have working verifier vagrant image with me.

I tested this in Windows Sandbox and locally and it works fine. Maybe verifier should be skipped now. Alternative would be to add vcredist2008 as dependency and hope it skip it.

**Comment by nutterthanos** [comment]
![image](https://user-images.githubusercontent.com/50470661/104796318-1aa47b80-5805-11eb-9d4a-8b695019b61c.png)
it got rejected

**Comment by nutterthanos** [comment]
> Is this still relevant? If so, what is blocking it? Is there anything you can do to help move it forward?

@AdmiringWorm

**Comment by nutterthanos** [comment]
this can be closed

**Comment by majkinetor** [comment]
I opened this issue. I don't have time to fix AHK script but anybody feel free to PR.

**Comment by jpluimers** [comment]
I figured out one thing already: https://github.com/chocolatey-community/chocolatey-coreteampackages/blob/master/automatic/nmap/tools/install.ahk has  `Npcap 1.00 Setup` everywhere, and nmap 7.92 tries to install npcap 1.50, so the dialogs are now titled `Npcap 1.50 Setup`.

I'm still recovering from extensive rectum cancer treatment, so if others have more energy than I: feel free to modify the autohotkey script.

**Comment by nutterthanos** [comment]
> Autohotkey does not make this a silent install:
> 
> ```
> You have nmap v7.80 installed. Version 7.92 is available based on your source(s).
> Progress: Downloading autohotkey.install 1.1.33.09... 100%
> Progress: Downloading nmap 7.92... 100%
> 
> autohotkey.install v1.1.33.09 [Approved]
> autohotkey.install package files upgrade completed. Performing other installation steps.
> Installing autohotkey.install...
> autohotkey.install has been installed.
> AutoHotkey* installed to 'C:\Program Files\AutoHotkey'
> AutoHotkey* registered as AutoHotkey
>   autohotkey.install can be automatically uninstalled.
>  The upgrade of autohotkey.install was successful.
>   Software installed as 'exe', install location is likely default.
> 
> nmap v7.92 [Approved]
> nmap package files upgrade completed. Performing other installation steps.
> Chocolatey timed out waiting for the command to finish. The timeout
>  specified (or the default value) was '2700' seconds. Perhaps try a
>  higher `--execution-timeout`? See `choco -h` for details.
>   nmap may be able to be automatically uninstalled.
> The upgrade of nmap was NOT successful.
> Error while running 'C:\ProgramData\chocolatey\lib\nmap\tools\chocolateyInstall.ps1'.
>  See log for details.
> This is try 1/3. Retrying after 300 milliseconds.
>  Error converted to warning:
>  (5) Access is denied: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap]
> This is try 2/3. Retrying after 400 milliseconds.
>  Error converted to warning:
>  (5) Access is denied: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap]
> Maximum tries of 3 reached. Throwing error.
> This is try 1/3. Retrying after 300 milliseconds.
>  Error converted to warning:
>  (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
> This is try 2/3. Retrying after 400 milliseconds.
>  Error converted to warning:
>  (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
> Maximum tries of 3 reached. Throwing error.
> Could not move bad package to failure directory It will show as installed.
>  C:\ProgramData\chocolatey\lib\nmap
>  The error:
>  (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
> This is try 1/3. Retrying after 300 milliseconds.
>  Error converted to warning:
>  (183) Cannot create a file when that file already exists: [\\?\C:\ProgramData\chocolatey\lib\nmap]
> This is try 2/3. Retrying after 400 milliseconds.
>  Error converted to warning:
>  (183) Cannot create a file when that file already exists: [\\?\C:\ProgramData\chocolatey\lib\nmap]
> Maximum tries of 3 reached. Throwing error.
> ```
> 
> Then a non-silent install aborted because I was away from keyboard during the update.
> 
> Should I re-open an issue for this?
> 
> --jeroen

I feel like this only happened to you since it didn't happen to me

**Comment by jpluimers** [comment]
> I feel like this only happened to you since it didn't happen to me

That's quite a binary observation for N=2 statistics (;

I managed to have a few "good" hours performing the monthly updates on all the systems here. Nmap failed consistently on the machines it was installed on (Windows 7, 8 and 10 VMs); all had this update to be applied:

`You have nmap v7.80 installed. Version 7.92 is available based on your source(s).`

All had a dialog like below waiting (I managed to take screenshots of a Windows 10 and a Windows 8.1 machine):

![nmap 7 92 hangs on npcap install during the license agreement - Screenshot 2021-08-12 095021](https://user-images.githubusercontent.com/2033367/129158956-191e7cb9-9ef3-4bd3-8402-fae3a70f587e.png)

![nmap 7 92 hangs on npcap install during the license agreement - Screenshot 2021-08-11 184346](https://user-images.githubusercontent.com/2033367/129158724-9a1433bb-bf5e-48df-ba90-4d42af57bded.png)

This is consistent with the npcap 1.00 -> 1.50 version upgrade I mentioned above in https://github.com/chocolatey-community/chocolatey-coreteampackages/issues/1557#issuecomment-897037373

**Comment by nutterthanos** [comment]
> > I feel like this only happened to you since it didn't happen to me
> 
> That's quite a binary observation for N=2 statistics (;
> 
> I managed to have a few "good" hours performing the monthly updates on all the systems here. Nmap failed consistently on the machines it was installed on (Windows 7, 8 and 10 VMs); all had this update to be applied:
> 
> `You have nmap v7.80 installed. Version 7.92 is available based on your source(s).`
> 
> All had a dialog like below waiting (I managed to take screenshots of a Windows 10 and a Windows 8.1 machine):
> 
> ![nmap 7 92 hangs on npcap install during the license agreement - Screenshot 2021-08-12 095021](https://user-images.githubusercontent.com/2033367/129158956-191e7cb9-9ef3-4bd3-8402-fae3a70f587e.png)
> 
> ![nmap 7 92 hangs on npcap install during the license agreement - Screenshot 2021-08-11 184346](https://user-images.githubusercontent.com/2033367/129158724-9a1433bb-bf5e-48df-ba90-4d42af57bded.png)
> 
> This is consistent with the npcap 1.00 -> 1.50 version upgrade I mentioned above in [#1557 (comment)](https://github.com/chocolatey-community/chocolatey-coreteampackages/issues/1557#issuecomment-897037373)
> 
> My logs are consistent with the failure at https://gist.github.com/choco-bot/b2c28eab3ed27d2de3f6393556270777#file-install-txt-L879 which odly is the base for "All Checks are Passing" at https://community.chocolatey.org/packages/nmap#testingResults
> 
> @majkinetor, I think the Choco-Bot should have failed the checks, so nmap 7.92 would have not been approved. Who can shed more insight on that?

this is pretty much what im getting ahk gets stuck

**Comment by pauby** [comment]
@nutterthanos Can you cut down on the amount you are quoting to only that which is relevant to your response?

**Comment by nutterthanos** [comment]
Ok

**Comment by majkinetor** [comment]
@jpluimers please check latest version: https://community.chocolatey.org/packages/nmap/7.92.0.20210815 using cinst with explicit version.

Not sure why the verification failed tho.

**Comment by jpluimers** [comment]
> @jpluimers please check latest version: https://community.chocolatey.org/packages/nmap/7.92.0.20210815 using cinst with explicit version.

I tried one failing system system today, and it works on that system; will try more systems later.

Note that the 7.80 install was really silent (no installation windows appeared); the 7.92 is half-silent: the dialogs show while autohotkey plays the right parts of the script.

Console log on a Windows 10 machine named `DESKTOP-4JJBH0K` (that's for my reference):

``` text
C:\bin\bin>nmap --version
Nmap version 7.80 ( https://nmap.org )
Platform: i686-pc-windows-windows
Compiled with: nmap-liblua-5.3.5 openssl-1.0.2s nmap-libssh2-1.8.2 nmap-libz-1.2.11 nmap-libpcre-7.6 Npcap-0.9982 nmap-libdnet-1.12 ipv6
Compiled without:
Available nsock engines: iocp poll select

C:\bin\bin>choco install --yes nmap --version 7.92.0.20210815
Chocolatey v0.10.15
Installing the following packages:
nmap
By installing you accept licenses for the packages.
Progress: Downloading nmap 7.92.0.20210815... 100%

nmap v7.92.0.20210815 - Possibly broken
nmap package files install completed. Performing other installation steps.
  nmap may be able to be automatically uninstalled.
 ShimGen has successfully created a shim for nmap-7.92-setup.exe
 The install of nmap was successful.
  Software install location not explicitly set, could be in package or
  default install location if installer.

Chocolatey installed 1/1 packages.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
 ```

> Not sure why the verification failed tho (@pauby please restart it or check it out). I tested this in Windows Sandbox and it installed correctly. RegEx in version should make this kind of bug not possible in the future.

I checked out the [verification](https://gist.github.com/choco-bot/d802589421b3cb5954c69eba7b2285be) and the [failure there](https://gist.github.com/choco-bot/d802589421b3cb5954c69eba7b2285be#file-install-txt-L784) is indeed odd; these files were successfully deleted on my system:

``` text
2021-08-19 17:28:56,380 16612 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815".
2021-08-19 17:28:56,399 16612 [DEBUG] - There was no original file at 'C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.registry'
2021-08-19 17:28:56,409 16612 [DEBUG] - There was no original file at 'C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.files'
2021-08-19 17:28:56,411 16612 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.extra".
2021-08-19 17:28:56,411 16612 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.version".
2021-08-19 17:28:56,412 16612 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.sxs".
2021-08-19 17:28:56,412 16612 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.pin".
2021-08-19 17:28:56,414 16612 [DEBUG] - Sending message 'HandlePackageResultCompletedMessage' out if there are subscribers...
2021-08-19 17:28:56,415 16612 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\lib\nmap\.chocolateyPending".
2021-08-19 17:28:56,418 16612 [INFO ] -  The install of nmap was successful.
```

**Comment by majkinetor** [comment]
> ShimGen has successfully created a shim for nmap-7.92-setup.exe

I will fix this and push it again, maybe verification will pass this time since @pauby didn't help (obviously appears here only for irrelevant stuff).

> Note that the 7.80 install was really silent (no installation windows appeared);

That can't be right. It also has AHK script which relies on window appearing, maybe they were not on top tho...

Anyway, it seems that this update fixes the problem as you confirmed. I checked installation in Windows Sandbox and it also works OK.

**Comment by jpluimers** [comment]
> > ShimGen has successfully created a shim for nmap-7.92-setup.exe
> 
> I will fix this and push it again, maybe verification will pass this time since @pauby didn't help (obviously appears here only for irrelevant stuff).

Hopefully.

In the mean time, it failed on a Windows 8.1 system, and I think it very much matches the [failure in the verification](https://gist.github.com/choco-bot/d802589421b3cb5954c69eba7b2285be#file-install-txt-L784).

First the console text, which already reveals a tiny bit:

``` text
C:\bin\bin>choco install --yes nmap --version 7.92.0.20210815
Chocolatey v0.10.15
Installing the following packages:
nmap
By installing you accept licenses for the packages.
Progress: Downloading nmap 7.92.0.20210815... 100%

nmap v7.92.0.20210815 - Possibly broken
nmap package files install completed. Performing other installation steps.
Chocolatey timed out waiting for the command to finish. The timeout
 specified (or the default value) was '2700' seconds. Perhaps try a
 higher `--execution-timeout`? See `choco -h` for details.
  nmap may be able to be automatically uninstalled.
The install of nmap was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\nmap\tools\chocolateyInstall.ps1'.
 See log for details.
This is try 1/3. Retrying after 300 milliseconds.
 Error converted to warning:
 (5) Access is denied: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap]
This is try 2/3. Retrying after 400 milliseconds.
 Error converted to warning:
 (5) Access is denied: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap]
Maximum tries of 3 reached. Throwing error.
This is try 1/3. Retrying after 300 milliseconds.
 Error converted to warning:
 (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
This is try 2/3. Retrying after 400 milliseconds.
 Error converted to warning:
 (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
Maximum tries of 3 reached. Throwing error.
Could not move bad package to failure directory It will show as installed.
 C:\ProgramData\chocolatey\lib\nmap
 The error:
 (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]

Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - nmap (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\nmap\tools\chocolateyInstall.ps1'.
 See log for details.
```

The log is far more similar to the [failure in the verification](https://gist.github.com/choco-bot/d802589421b3cb5954c69eba7b2285be#file-install-txt-L784):

``` text
2021-08-19 19:11:32,640 4036 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815".
2021-08-19 19:11:32,684 4036 [DEBUG] - There was no original file at 'C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.registry'
2021-08-19 19:11:32,714 4036 [DEBUG] - There was no original file at 'C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.files'
2021-08-19 19:11:32,717 4036 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.arguments".
2021-08-19 19:11:32,719 4036 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.extra".
2021-08-19 19:11:32,720 4036 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.version".
2021-08-19 19:11:32,722 4036 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.sxs".
2021-08-19 19:11:32,723 4036 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\nmap.7.92.0.20210815\.pin".
2021-08-19 19:11:32,728 4036 [DEBUG] - Sending message 'HandlePackageResultCompletedMessage' out if there are subscribers...
2021-08-19 19:11:32,733 4036 [ERROR] - The install of nmap was NOT successful.
2021-08-19 19:11:32,738 4036 [ERROR] - Error while running 'C:\ProgramData\chocolatey\lib\nmap\tools\chocolateyInstall.ps1'.
 See log for details.
2021-08-19 19:11:32,741 4036 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\lib-bad".
2021-08-19 19:11:32,749 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap'
2021-08-19 19:11:32,843 4036 [WARN ] - This is try 1/3. Retrying after 300 milliseconds.
 Error converted to warning:
 (5) Access is denied: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap]
2021-08-19 19:11:33,150 4036 [WARN ] - This is try 2/3. Retrying after 400 milliseconds.
 Error converted to warning:
 (5) Access is denied: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap]
2021-08-19 19:11:33,570 4036 [ERROR] - Maximum tries of 3 reached. Throwing error.
2021-08-19 19:11:33,574 4036 [WARN ] - Move failed with message:
 (5) Access is denied: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap]
 Attempting backup move method.
2021-08-19 19:11:33,576 4036 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\lib-bad\nmap".
2021-08-19 19:11:33,578 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\.chocolateyPending'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\.chocolateyPending'
2021-08-19 19:11:33,582 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\nmap.nupkg'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\nmap.nupkg'
2021-08-19 19:11:33,585 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\nmap.nuspec'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\nmap.nuspec'
2021-08-19 19:11:33,587 4036 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\lib-bad\nmap\legal".
2021-08-19 19:11:33,590 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\legal\LICENSE.txt'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\legal\LICENSE.txt'
2021-08-19 19:11:33,592 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\legal\VERIFICATION.txt'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\legal\VERIFICATION.txt'
2021-08-19 19:11:33,596 4036 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\lib-bad\nmap\tools".
2021-08-19 19:11:33,598 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\tools\chocolateyInstall.ps1'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\tools\chocolateyInstall.ps1'
2021-08-19 19:11:33,601 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\tools\chocolateyUninstall.ps1'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\tools\chocolateyUninstall.ps1'
2021-08-19 19:11:33,603 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\tools\install.ahk'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\tools\install.ahk'
2021-08-19 19:11:33,607 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\tools\nmap-7.80-setup.exe.ignore'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.80-setup.exe.ignore'
2021-08-19 19:11:33,611 4036 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\nmap\tools\nmap-7.92-setup.exe'
 to 'C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe'
2021-08-19 19:11:33,625 4036 [WARN ] - This is try 1/3. Retrying after 300 milliseconds.
 Error converted to warning:
 (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
2021-08-19 19:11:33,932 4036 [WARN ] - This is try 2/3. Retrying after 400 milliseconds.
 Error converted to warning:
 (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
2021-08-19 19:11:34,339 4036 [ERROR] - Maximum tries of 3 reached. Throwing error.
2021-08-19 19:11:34,342 4036 [ERROR] - Could not move bad package to failure directory It will show as installed.
 C:\ProgramData\chocolatey\lib\nmap
 The error:
 (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\nmap\tools\nmap-7.92-setup.exe]
2021-08-19 19:11:34,354 4036 [DEBUG] - Attempting to delete file "C:\Users\jeroenp\AppData\Local\NuGet\Cache\nmap.7.92.0.20210815.nupkg".
2021-08-19 19:11:34,384 4036 [WARN ] - 
Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
```

The really odd thing is the nmap installation dialog that was still visible: the Finished button was highlighted, but the [only AHK script bit having a `break` out of the `loop`](https://github.com/chocolatey-community/chocolatey-coreteampackages/blob/60684a02abee52af97e76c670fba30d89a6ab269/automatic/nmap/tools/install.ahk#L36) missed it:

``` autohotkey
    IfWinExist, Nmap Setup, Installation Complete
    {
        Sleep, 100
        ControlSend,, {Enter}
        Sleep, 100
        ControlSend,, {Enter}
        Sleep, 100
        ControlSend,, {Enter}
        Sleep, 100
        ControlSend,, {Enter}
        break
    }
```

The dialog is this:

![nmap 7 92 hangs during final installation step - Screenshot 2021-08-19 182821](https://user-images.githubusercontent.com/2033367/130126536-26bd2b8f-c7ae-41d1-9c27-65f65015a895.png)

Pressing the `Enter` key of course terminated the dialog.

Though I left the machine alone during installation (had dinner) to prevent any spurious focus issues, I have been bitten by user-interface automation automation issues back in the Microsoft Active Accessibility days). Gut feeling it's either a timing issue somewhere, or maybe a dialog text that is different: I could not find `Installation Complete` anywhere in that dialog.

> > Note that the 7.80 install was really silent (no installation windows appeared);

> That can't be right. It also has AHK script which relies on window appearing, maybe they were not on top tho...

That thought went to my mind as well, and you are almost right. Had to dig quite a bit (I really dislike commits in this repositories that update more than one tool when the actual changes are unrelated), but found it:

- 7.80 was still a silent install and did not have AHK in the tools directory https://github.com/chocolatey-community/chocolatey-coreteampackages/tree/dff56f5203ce947916524e72d64cd9d504d9c500/automatic/nmap/tools
- 7.90 introduced AHK: https://github.com/chocolatey-community/chocolatey-coreteampackages/blob/842ea37d69474c75ae81afbe5dcc0306681fd562/automatic/nmap/tools/install.ahk
- this matches the Discus comment here: https://community.chocolatey.org/packages/nmap#comment-5095392481

> Anyway, it seems that this update fixes the problem as you confirmed. I checked installation in Windows Sandbox and it also works OK.

I'll test more systems, just to see if that nmap install dialog hangs on others as well and if so: if there is a pattern. My gut feeling has failed me some of the times, so more investigation is needed (:

**Comment by majkinetor** [comment]
> Though I left the machine alone during installation (had dinner) to prevent any spurious focus issues,

I think focus shouldn't make an issue with this AHK script as it targets windows directly and where not it should use last detected previously.

On Sandbox script finishes reliably each time I tried.  

> 7.80 was still a silent install 

Ah, thx for reminder, that definitely the reason then. 

> I'll test more systems, just to see if that nmap install dialog hangs on others as well and if so: if there is a pattern. My gut feeling has failed me some of the times, so more investigation is needed (:

OK, check it out, and let me know. Maybe I can improve AHK script a bit but not by much. GUI automation is often not reliable and I rarely use AHK nowdays.

**Comment by jpluimers** [comment]
> > I'll test more systems, just to see if that nmap install dialog hangs on others as well and if so: if there is a pattern. My gut feeling has failed me some of the times, so more investigation is needed (:
> 
> OK, check it out, and let me know. Maybe I can improve AHK script a bit but not by much. GUI automation is often not reliable and I rarely use AHK nowdays.

Will do. One thing in the AHK script that strikes me is the absolute lack of documentation: which control is supposed to have focus when the `Enter` key is pressed.

One thought is to run the downloaded installer manually on a succeeding system and on a failing system, then by hand follow the steps in the AHK script and document it in a gist so you can later use those.

For my link archive: https://www.autohotkey.com/docs/Scripts.htm#Comments

**Comment by majkinetor** [comment]
> which control is supposed to have focus when the Enter key is pressed.
In this case, last found window: https://www.autohotkey.com/docs//misc/WinTitle.htm#LastFoundWindow

**Comment by jpluimers** [comment]
> > which control is supposed to have focus when the Enter key is pressed.
> 
> In this case, last found window: https://www.autohotkey.com/docs//misc/WinTitle.htm#LastFoundWindow

Thanks for that.

In the mean time, the only systems that fail consistently are Windows 8.1 based ones. Windows 7 and 10 work fine. Don't have Windows 11 yet.

To be continued.

**Comment by majkinetor** [comment]
In the meantime, the verification still fails. /cc @AdmiringWorm

**Comment by jpluimers** [comment]
> In the meantime, the verification still fails. /cc @AdmiringWorm

The cause for the failure might be the same as on my Windows 8.1 systems.

Will take me longer to take a look into that, as the mom of my best friend is not doing well, so he moved from abroad to live with us for a while.

**Comment by majkinetor** [comment]
> The cause for the failure might be the same as on my Windows 8.1 systems.

Must be. Verification server runs on Windows Server 2012 which AFAIK is 8.1 era

> Will take me longer to take a look into that, as the mom of my best friend is not doing well, so he moved from abroad to live with us for a while.

Too much detail my man :)

**Comment by jpluimers** [comment]
> > The cause for the failure might be the same as on my Windows 8.1 systems.
> 
> Must be. Verification server runs on Windows Server 2012 which AFAIK is 8.1 era

Indeed. Did some more work, and the only Windows 10 machine it failed on, failed only once, so was not reproducible. All reproducible cases are Windows 8.1 virtual machines.

(Sorry for the extra details).

Hopefully I can play around with either that or the Chocolatey Testing Environment so the package verifier won't barf on it any more.

For my memory:

- https://docs.chocolatey.org/en-us/community-repository/moderation/package-verifier
- https://github.com/chocolatey-community/chocolatey-test-environment

**Comment by ottercynical** [comment]
Still a bug in 2025.

**Comment by corbob** [comment]
> Still a bug in 2025.

Feel free to open a new issue with details as this issue has been closed for 3 years @ottercynical .
